adapters/src/main/java/com/squareup/moshi/adapters/RuntimeJsonAdapterFactory.java

```diff
@@ -140,6 +150,25 @@
     }
 
     @Override public void toJson(JsonWriter writer, Object value) throws IOException {
+      if (!baseTypeIsObject) {
+        objectJsonAdapter.toJson(writer, toJsonMap(value));
+        return;
+      }
+      if (reentrantFromObjectAdapter) {
```

Comment:
I don't think this works for recursive structures like trees.

Reply:
Also this isn't threadsafe.

Reply:
Curious: how does it fail without this?

Reply:
ahaha, yeah, of course.

>how does it fail without this?

the test in #664 fails on master. is that what you mean?